### PR TITLE
Fix master build with --strict-warnings.

### PR DESCRIPTION
**Pull request by FdaSilvaYY**
Found while trying to a CI failure.

~~Improve `nmake clean` cleanup.~~
 

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by FdaSilvaYY** [comment]
Ping @DDvO  : feel free to keep my fix or discard all the `const` in this method.

File: crypto/cmp/cmp_ctx.c [hunk 1]
```diff
@@ -425,7 +425,7 @@ int OSSL_CMP_CTX_set1_referenceValue(OSSL_CMP_CTX *ctx,
 
 /* Set or clear the password to be used for protecting messages with PBMAC */
 int OSSL_CMP_CTX_set1_secretValue(OSSL_CMP_CTX *ctx, const unsigned char *sec,
-                                  int len)
+                                  const int len)
```
Comment (mspncp) [inline crypto/cmp/cmp_ctx.c:428]: What is the point of marking the parameter `len` as `const` in the first place? It's passed by value, so the caller does need to care whether the value is modified inside the function.

In my opinion, it's not the function signature of the implementation which needs to be changed, it's the signature of the prototype in the header and the documentation.

```
~/src/openssl/master$ git grep -n -A 1 OSSL_CMP_CTX_set1_secretValue | grep -B 1 'int len'
crypto/cmp/cmp_ctx.c:421:int OSSL_CMP_CTX_set1_secretValue(OSSL_CMP_CTX *ctx, const unsigned char *sec,
crypto/cmp/cmp_ctx.c-422-                                  int len)
--
doc/man3/OSSL_CMP_CTX_new.pod:118: int OSSL_CMP_CTX_set1_secretValue(OSSL_CMP_CTX *ctx, const unsigned char *sec,
doc/man3/OSSL_CMP_CTX_new.pod-119-                                   const int len);
--
include/openssl/cmp.h.in:327:int OSSL_CMP_CTX_set1_secretValue(OSSL_CMP_CTX *ctx, const unsigned char *sec,
include/openssl/cmp.h.in-328-                                  const int len);
```

**Review by mspncp — CHANGES_REQUESTED** [review]

File: crypto/cmp/cmp_ctx.c (hunk 1 above)
Comment (FdaSilvaYY) [reply crypto/cmp/cmp_ctx.c:428]: That's why I ping @DDvO to get its feeling, because this way of fixing it is much larger.
This could be a merge conflict remainder or a review issue.

Comment (levitte) [reply crypto/cmp/cmp_ctx.c:428]: constifying scalar function parameters is a bit over-zealous, I agree.
I think that in certain places, this is done to indicate an intent that the input parameter will be used unchanged, i.e. will not be "adjusted" frivolously.

Comment (mspncp) [reply crypto/cmp/cmp_ctx.c:428]: In other words: it's leaking implementation details without good cause ;-)

Comment (levitte) [reply crypto/cmp/cmp_ctx.c:428]: > In other words: it's leaking implementation details without good cause ;-)

No, it's making a promise.

File: Configurations/windows-makefile.tmpl [hunk 2]
```diff
@@ -471,11 +471,12 @@ clean: libclean
 	{- join("\n\t", map { "-del /Q /F $_" } @SCRIPTS) || "\@rem" -}
 	{- join("\n\t", map { "-del /Q /F $_" } @GENERATED_MANDATORY) || "\@rem" -}
 	{- join("\n\t", map { "-del /Q /F $_" } @GENERATED) || "\@rem" -}
-	-del /Q /S /F *.d *.obj *.pdb *.ilk *.manifest
+	-del /Q /S /F *.d *.obj *.pdb *.ilk *.manifest *.asm
```
Comment (levitte) [inline Configurations/windows-makefile.tmpl:474]: Aren't the generated .asm files already part of `@GENERATED` and therefore removed in line 473?

File: Configurations/windows-makefile.tmpl [hunk 3]
```diff
@@ -471,11 +471,12 @@ clean: libclean
...
 	-del /Q /S /F engines\*.lib engines\*.exp
 	-del /Q /S /F apps\*.lib apps\*.rc apps\*.res apps\*.exp
-	-del /Q /S /F test\*.exp
-	-rd /Q /S test\test-runs
+	-del /Q /S /F test\*.exp test\*.d test\*.obj test\*.pdb test\*.exe
```
Comment (levitte) [inline Configurations/windows-makefile.tmpl:477]: All of these except `test\*.exp` are already removed in line 474...

File: Configurations/windows-makefile.tmpl [hunk 4]
```diff
@@ -471,11 +471,12 @@ clean: libclean
...
+	-del /Q /S /F test\buildtest_*.c
```
Comment (levitte) [inline Configurations/windows-makefile.tmpl:478]: Aren't these already part of `@GENERATED` and therefore removed in line 473?

File: Configurations/windows-makefile.tmpl [hunk 5]
```diff
@@ -471,11 +471,12 @@ clean: libclean
...
+	-rd /Q /S test\test-runs test-runs
```
Comment (levitte) [inline Configurations/windows-makefile.tmpl:479]: Actually, `test\test-runs` should be dropped.

**Review by levitte — CHANGES_REQUESTED** [review]

File: crypto/cmp/cmp_ctx.c (hunk 1 above)
Comment (DDvO) [reply crypto/cmp/cmp_ctx.c:428]: Sorry for my somewhat delayed response, I did not notice this PR and question to me before.

I cannot recall how that `const int` came in originally and see little practical use for it
(except that it may help the implementer of the function not to accidentally modify the parameter value).

As one can easily find out from the commit history, already in commit 08dfbe0798f57ac9e9793fdfcaff54cfdf6b3359 I removed the `const`,
so please do not re-introduce it here.
BTW, @FdaSilvaYY I just found that you had commented on that PR, though on a different topic: https://github.com/openssl/openssl/pull/17284#issuecomment-997202727.

Unfortunately, `cmp.h.in` still contains
```
int OSSL_CMP_CTX_set1_secretValue(OSSL_CMP_CTX *ctx, const unsigned char *sec,
                                  const int len);
```
so I presume that you aim to make the function definition in `cmp_ctx.c` consistent with that.

I'd much prefer if the `const` instead gets removed from the declaration in the header file,
yet we then face the problem that this appears (syntactically) to be an API change/break,
while actually (semantically) it has no effect on the API (i.e., the interface and behavior for callers).

Comment (FdaSilvaYY) [reply crypto/cmp/cmp_ctx.c]: @DDvO I push a fix in this way.

**Review by DDvO — APPROVED** [review]
Thanks for the cleanup

**Review by DDvO — CHANGES_REQUESTED** [review]
Ah, one more thing: please change the `const int` also in `OSSL_CMP_CTX_new.pod`.

A pity that `doc-nits` cannot check the consistency of (function etc.) type decls with the respective header files.

**Comment by DDvO** [comment]
Regarding the consistency of the documentation, the checklist in the PR text pattern is a nice hint:


> ##### Checklist
> 
> - [ ] documentation is added or updated
> - [X] tests are added or updated

**Comment by FdaSilvaYY** [comment]
> Ah, one more thing: please change the `const int` also in `OSSL_CMP_CTX_new.pod`.
> 
> A pity that `doc-nits` cannot check the consistency of (function etc.) type decls with the respective header files.

`OSSL_CMP_CTX_new.pod` fix is in my commit.

This PR is ready to review.

**Review by DDvO — APPROVED** [review]

**Comment by DDvO** [comment]
> > Ah, one more thing: please change the `const int` also in `OSSL_CMP_CTX_new.pod`.
> > A pity that `doc-nits` cannot check the consistency of (function etc.) type decls with the respective header files.
> 
> `OSSL_CMP_CTX_new.pod` fix is in my commit.
> 
> This PR is ready to review.

Oh, I had overlooked this - sorry for my confusion!
So I've renewed my approval.

File: include/openssl/cmp.h.in [hunk 6]
```diff
@@ -329,8 +329,8 @@ int OSSL_CMP_CTX_build_cert_chain(OSSL_CMP_CTX *ctx, X509_STORE *own_trusted,
 int OSSL_CMP_CTX_set1_pkey(OSSL_CMP_CTX *ctx, EVP_PKEY *pkey);
 int OSSL_CMP_CTX_set1_referenceValue(OSSL_CMP_CTX *ctx,
                                      const unsigned char *ref, int len);
-int OSSL_CMP_CTX_set1_secretValue(OSSL_CMP_CTX *ctx, const unsigned char *sec,
-                                  const int len);
+int OSSL_CMP_CTX_set1_secretValue(OSSL_CMP_CTX *ctx,
+                                  const unsigned char *sec, int len);
```
Comment (mspncp) [inline include/openssl/cmp.h.in:333]: It's a bit unfortunate that your whitespace reformatting hides the aktual fix, namely that the `const` was removed.

I suggest you create two commits (in this order, to facilitate `git blame`):

-  one commit for the typo and the whitespace changes
-  one commit for fixing the compiler-warning

File: crypto/cmp/cmp_ctx.c [hunk 7]
```diff
@@ -424,8 +424,8 @@ int OSSL_CMP_CTX_set1_referenceValue(OSSL_CMP_CTX *ctx,
 }
 
 /* Set or clear the password to be used for protecting messages with PBMAC */
-int OSSL_CMP_CTX_set1_secretValue(OSSL_CMP_CTX *ctx, const unsigned char *sec,
-                                  int len)
+int OSSL_CMP_CTX_set1_secretValue(OSSL_CMP_CTX *ctx,
+                                  const unsigned char *sec, int len)
```
Comment (mspncp) [inline crypto/cmp/cmp_ctx.c:428]: This is an unrelated and distracting whitespace change. If you want to keep it, please split it into a separate commit.

File: doc/man3/OSSL_CMP_CTX_new.pod [hunk 8]
```diff
@@ -119,8 +119,8 @@ OSSL_CMP_CTX_set1_senderNonce
  int OSSL_CMP_CTX_set1_pkey(OSSL_CMP_CTX *ctx, EVP_PKEY *pkey);
  int OSSL_CMP_CTX_set1_referenceValue(OSSL_CMP_CTX *ctx,
                                       const unsigned char *ref, int len);
- int OSSL_CMP_CTX_set1_secretValue(OSSL_CMP_CTX *ctx, const unsigned char *sec,
-                                   const int len);
+ int OSSL_CMP_CTX_set1_secretValue(OSSL_CMP_CTX *ctx,
+                                   const unsigned char *sec, int len);
```
Comment (mspncp) [inline doc/man3/OSSL_CMP_CTX_new.pod:123]: dito

**Review by mspncp — CHANGES_REQUESTED** [review]

**Comment by FdaSilvaYY** [comment]
@mspncp : changes split in two commits
Please review

**Review by DDvO — APPROVED** [review]
reconfirmed

**Comment by mspncp** [comment]
Last [force-push](https://github.com/openssl/openssl/compare/003db5cab0dc363bf7c00c29f86a0783cdff27df..c7214af1e01e90ce5f8561c048b602aef55d58ac) was without tree changes.

@FdaSilvaYY sorry, I couldn't resist:  I swapped the two commits: the whitespace reformatting needs to go _first_ (in topological order), otherwise it covers the relevant change in the annotation. Also, I inserted an an empty line between subject line and body in one of the commit messages.

**Review by mspncp — APPROVED** [review]

**Comment by FdaSilvaYY** [comment]
PR ready for merge ?
Did it deserve a backport ?

**Comment by DDvO** [comment]
> PR ready for merge ?

Yes, but the repo is currently frozen, so needs to wait until tomorrow.

> Did it deserve a backport ?

I'd say yes as it may help avoiding build hickups.

**Comment by t8m** [comment]
@FdaSilvaYY where do you see the warning, which compiler/OS?

**Comment by mattcaswell** [comment]
There is a technical API break here due to dropping the `const` from `OSSL_CMP_CTX_set1_secretValue` in the header. Needs an OTC decision.

**Comment by mspncp** [comment]
The OTC discussed several alternative options and did not reach a consensus yet. We will resume the discussion and vote on it next week. Since it is an API change, by our policies it would be allowed only for the next major release, unless we vote for an exception. The controversial question is whether it should be fixed in the header or in the source in all of 3.x, or some mixed solution (fix the header in 'master' branch, backport a source file fix only).

**Comment by hlandau** [comment]
My own position is:
- I don't care whether we fix the source or fix the header.
- I would `[-1]` any option which means we adopt different fixes for different branches.
- I would `[-1]` any option which means we don't fix all of 3.0, 3.1, and master.

**Comment by DDvO** [comment]
As I wrote 4 days ago, above in https://github.com/openssl/openssl/pull/20436#discussion_r1131084897,
removing the `const` on a function parameter of scalar type only **appears** to be an API break.

> I'd much prefer if the const instead gets removed from the declaration in the header file,
yet we then face the problem that this appears (syntactically) to be an API change/break,
while actually (semantically) it has no effect on the API (i.e., the interface and behavior for callers).

**Comment by mspncp** [comment]
> removing the const on a function parameter of scalar type only appears to be an API break.

Are you really sure your assertion includes the (unlikely) case where some user has typedef'ed a "pointer to this function" type for assignment, on every supported compiler?

**Comment by levitte** [comment]
> > removing the const on a function parameter of scalar type only appears to be an API break.
> 
> Are you really sure your assertion includes the (unlikely) case where some user has typedef'ed a "pointer to this function" type for assignment, on every supported compiler?

(nit: no typedef required, all that's needed is that someone expects a pointer matching the current function signature)

I think that's the point being discussed, i.e. how likely is it that this API change breaks anything, or in other words, how much do we have to care.  The semanticaly miniscule difference between a const or non-const argument passed by value makes no difference for someone who's build of their application suddenly breaks.

**Comment by DDvO** [comment]
> > removing the const on a function parameter of scalar type only appears to be an API break.
> 
> Are you really sure your assertion includes the (unlikely) case where some user has typedef'ed a "pointer to this function" type for assignment, on every supported compiler?

Yes.
Regardless whether you use a function directly or via a function pointer,
for the caller it does not matter whether the function implementation modifies a parameter with a primitive (i.e., not reference) type, 
simply because the value is anyway passed by copying.

**Comment by DDvO** [comment]
I suggest clarifying this once and for all, not only for this PR,
and sharpening the OpenSSL definition/understanding of an API break accordingly.

BTW, where can this definition be found?

**Comment by levitte** [comment]
Personally, I've no issue with changing the header, as I believe (without any data to confirm) very few if any at all have reason to pass a pointer to that particular function.  But, I'm with @hlandau on this, that if we fix the header, it should be done consistently for all 3.x releases, even 3.0.

**Comment by mattcaswell** [comment]
I've just run a test, attempting to compile with this code:

```` C
#include <openssl/cmp.h>

int main(void)
{
    int (*secretValue)(OSSL_CMP_CTX *ctx, const unsigned char *sec,
                    const int len);

    secretValue = OSSL_CMP_CTX_set1_secretValue;

    secretValue(NULL, NULL, 0);
    return 0;
}
````
Compiled with `-Wall` using both gcc (11.3) and clang (14.0) and against a version of OpenSSL which has dropped the `const` for the `len` parameter from the header. I was unable to get either to issue a warning about this. Not a guarantee for all compilers of course.

**Comment by beldmit** [comment]
What about abidiff?

**Comment by DDvO** [comment]
Again, whether a function pointer is used does not make any difference here.

It just might happen that some compiler does not correctly reflect that on matching a primitive function parameter type (such as `int`) it does not matter whether it is preceded by `const` or not.
Glad to see that at least current `gcc` and `clang` have no issue here.

**Comment by levitte** [comment]
> I've just run a test, attempting to compile with this code:
> 
> ```c
> #include <openssl/cmp.h>
> 
> int main(void)
> {
>     int (*secretValue)(OSSL_CMP_CTX *ctx, const unsigned char *sec,
>                     const int len);
> 
>     secretValue = OSSL_CMP_CTX_set1_secretValue;
> 
>     secretValue(NULL, NULL, 0);
>     return 0;
> }
> ```
> 
> Compiled with `-Wall` using both gcc (11.3) and clang (14.0) and against a version of OpenSSL which has dropped the `const` for the `len` parameter from the header. I was unable to get either to issue a warning about this. Not a guarantee for all compilers of course.

This search indicates that clang 15 is currently packaged in Debian [testing], which will become the next stable release:

https://packages.debian.org/search?keywords=clang

Looking at Debians release history and cadence (https://wiki.debian.org/DebianReleases), I would expect Debian 12 (which is what Debian [testing] is right now) to be released some time this year...
Disclaimer: I'm an not a Debian developer / maintainer

**Comment by levitte** [comment]
> Glad to see that at least current `gcc` and `clang` have no issue here.

The latest releases for these are gcc 12 and clang 15, according to their own release pages:

https://gcc.gnu.org/releases.html
https://releases.llvm.org/

I can try with those

**Comment by levitte** [comment]
No problem with gcc-12 or clang-15 either:

``` C
/* I did it a little differently for demonstration purposes */
#include <stddef.h>
/* #include <openssl/cmp.h> */
typedef struct ossl_cmp_ctx_st OSSL_CMP_CTX;
int OSSL_CMP_CTX_set1_secretValue(OSSL_CMP_CTX *ctx, const unsigned char *sec,
                                  int len);

int main(void)
{
    int (*secretValue)(OSSL_CMP_CTX *ctx, const unsigned char *sec,
                       const int len);

    secretValue = OSSL_CMP_CTX_set1_secretValue;

    secretValue(NULL, NULL, 0);
    return 0;
}
```

``` console
$ gcc-12 -Wall -o 20436.o -c 20436.c
$ clang-15 -Wall -o 20436.o -c 20436.c
$
```

This confirms the "not a problem" standing as far as I can take it.
...  and this has me think that we should really fix the header, and do it all the way to 3.0.

**Comment by mattcaswell** [comment]
Could someone try with VisualStudio? If its not a problem there either then I'm also starting to lean towards "do it all the way to 3.0".

**Comment by hlandau** [comment]
It is not a problem on clang 15 or GCC or MSVC.

**Comment by hlandau** [comment]
C11 s. 6.7.6.3. p. 15. "(In the determination of type compatibility and of a composite type, each parameter [...] declared with qualified type is taken as having the unqualified version of its declared type.)"

**Comment by hlandau** [comment]
Looks like there should be no issue doing the header change, e.g. in 3.0, both in theory and in practice.

**Comment by DDvO** [comment]
> I suggest clarifying this once and for all, not only for this PR, and sharpening the OpenSSL definition/understanding of an API break accordingly.
> 
> BTW, where can this definition be found?

Is there a clear official OpenSSL definition of 
what exactly constitutes an API break, and how to determine this syntactically?

**Comment by FdaSilvaYY** [comment]
> @FdaSilvaYY where do you see the warning, which compiler/OS?

I was using Visual 2015, with --all-warnings, to debug a build break of my another PR.

**Comment by DDvO** [comment]
> > @FdaSilvaYY where do you see the warning, which compiler/OS?
> 
> I was using Visual 2015, with --strict-warnings, to debug a build break of my another PR.

Ah yeah, MickeySoft always tends to make trouble, with all sorts of incompatibility.

**Comment by paulidale** [comment]
Fix the header and merge back to 3.0 seems to be the consensus the group is coming to.
I can support this.

**Comment by t8m** [comment]
**OTC: Given the compilers ignore that const we want to fix the header and backport that to 3.1 and 3.0 branches.**

**Comment by t8m** [comment]
If this does not apply cleanly we will need a backport patch for 3.1.

**Comment by DDvO** [comment]
@openssl/otc, again my questions:

> > I suggest clarifying this once and for all, not only for this PR, and sharpening the OpenSSL definition/understanding of an API break accordingly.
> > BTW, where can this defin
ition be found?
> 
> Is there a clear official OpenSSL definition of what exactly constitutes an API break, and how to determine this syntactically?

At least the outcome of our above exchange should be remembered somehow
such that on any future such cases it does not need to get re-discovered with effort.

**Comment by t8m** [comment]
The problem is that basically anything that changes in the public API headers that can potential break a compilation for anyone is an API break. So until we found that this is very unlikely to break anyone it was considered an API break. Not sure how to record this and where.

**Comment by DDvO** [comment]
> The problem is that basically anything that changes in the public API headers that can potential break a compilation for anyone is an API break. So until we found that this is very unlikely to break anyone it was considered an API break. Not sure how to record this and where.

Thanks for your response, which more or less includes a definition of an API break.
Unfortunately, it won't be easy to give simple syntactic rules to correctly and completely determine what will constitute an API break or not.

For instance, the following header file changes do not lead to API breaks:
* whitespace and comment changes
* adding or removing the `const` qualifier before a primitive (i.e., non-reference) type
* Adding or removing the name or changing the name of a function parameter
* ...?

I suggest adding something like this to the technical policies, next to the coding style.

**Comment by t8m** [comment]
Added https://github.com/openssl/technical-policies/issues/65
